[Poker game app step5] - 포커게임 결과 화면 구현 진행중

dev-Lena opened the pull request:
### 현재 진행 상태

1. 게임모드를 선택할 수 있는 UISegmentedControl 추가했습니다.
2. 플레이어 수를  선택할 수 있는 UISegmentedControl 추가했습니다.
3. 하지만 UISegmentedControl 선택에 따른 액션이나 동작은 아직 안됩니다.
4.  현재 7cardsStud 와 플레이어 4명으로 설정한 상태입니다. 
4-1. 그런데 한 플레이어에 7장의 카드가 들어가고 있지 않습니다.
4-2. 또한 4명의 플레이어와 딜러 모두 같은 카드를 갖고 있습니다.
4-3. 7장의 카드가 수평으로 나란히 들어가지 않고 중간에 공백이 있습니다.

스냅샷 :
![](https://i.imgur.com/WuK2Yxl.png)


### 해결하지 못한 부분  
시도해봤지만 어디가 어떻게 문제인지 파악하지 못했습니다.. jk의 조언과 피드백이 필요해요...

1. 현재 7cardsStud 와 플레이어 4명으로 설정한 상태입니다. 하지만
1-1. 그런데 한 플레이어에 7장의 카드가 들어가고 있지 않습니다.
1-2. 또한 4명의 플레이어와 딜러 모두 같은 카드를 갖고 있습니다.

2. 아직 게임모드나 플레이어 수를 선택할 수 있는 SegmentedControl 선택에 따라 다른 게임판이 세팅되지 않습니다. 앞서 1번 문제를 해결 후 추가 구현할 예정입니다.

### 고민한 점

1. private으로 되어있는 플레이어나 딜러가 갖고 있는 카드의 설명(description)을 ViewController에서 사용해야하는데 어떻게 가져(?)올 수 있을지
→ 그 value를 가지고 있는 부분에서 forEach문을 이용한 메소드를 추가했습니다.
→ 클로저를 사용해봤습니다.

2. StackView 안에 StackView가 왜 들어가지 않을까
: 플레이어가 추가될 때 마다 StackView를 새로생성하지 않고 원래 있던 StackView를 계속 바꾸고 있었기 때문이었다는 걸 발견하고 수정했습니다.

3. stackView 구조
→ 전체 gameStackView 안에 PlayersStackView를 추가했습니다.
→ playersStackView에는 참가자 이름(Player1, Dealer...)의 UILabel과 참가자 카드들의 UIImageView들을 추가했습니다.
→ playersStackView에서 참가자 카드들의 UIImageView들은 cardsStackView에 추가했습니다.

dev-Lena committed:
불필요한 파일 삭제

dev-Lena committed:
Players 클래스의 addPlayer 메소드 수정 : forEach문 삭제

dev-Lena committed:
NumbersOfPlayers enum과 GameMode enum을 PokerGame클래스 외부로 분리

dev-Lena committed:
Players 클래스에 checkCardsFull()메소드 추가 및 PokerGame 클래스에 isAllPlayersCardsReady() 메소드 변경

dev-Lena committed:
Model Group 생성 후 Card,CardDeck,PokerGame,Participant,Players 클래스 추가

dev-Lena committed:
Assets에 카드 앞면 이미지 추가

dev-Lena committed:
gameMode(5 cards 또는 7 cards)를 선택하는 UISegmentedControl 추가

dev-Lena committed:
ViewController 에서 불필요한 메소드 삭제

dev-Lena committed:
Participant 클래스와 PokerGame의 init 메소드 수정

dev-Lena committed:
Dealer가 가지고 있는 CardDeck을 섞는 메소드 추가

dev-Lena committed:
Player의 Label을 만드는 메소드 추가

dev-Lena committed:
StackView 세팅하는 setStackView 메소드 수정

dev-Lena committed:
Player의 카드 이미지를 만들어 추가하는 makePlayersCards 메소드 추가

dev-Lena committed:
PokerGame에 참가한 Player들이 가지고 있는 Card 배열에 접근할 수 있도록 메소드 추가

dev-Lena committed:
PokerGame 클래스에서 distributeCards() 메소드 수정

dev-Lena committed:
플레이어들과 딜러들을 담을 GameStackView 추가

dev-Lena committed:
Dealer 라벨과 Dealer의 카드 이미지 playersStackView에 추가

dev-Lena committed:
Dealer와 Player들을 gameStackView에 추가

dev-Lena committed:
불필요한 코드 삭제

godrm commented on CardGameApp/CardGameApp/Model/PokerGame.swift:
```diff
@@ -0,0 +1,79 @@
+    // 플레이어의 자리가 몇 개가 필요한지 세팅
+    func setPlayerSeat(for bringPlayer: () -> ()) {
+        for _ in 1 ... self.rawValue {
+            bringPlayer()
```
주석으로 설명하려고 하기 보다는 코드를 읽고 이해할 수 있으면 됩니다.
클로저 내용을 모르는 데 클로저 이름을 bringPlayer 붙여준 것은 좋은데 사용하는 데 제한이 생기는 것 같은 느낌이네요. 
NumbersOfPlayers를 활용해서 반복하는 경우는 다 쓸 수 있지 않을까요? 몇 명이 필요한지만 사용하지 않을 것 같아요.
이 부분도 고려해보세요.

godrm commented on CardGameApp/CardGameApp/Model/PokerGame.swift:
```diff
@@ -0,0 +1,79 @@
+                players.addPlayer(newPlayer: Player())
+            })
+            return self.players
+        }()
```
클로저를 만들어서 동작을 전달하는 것은 좋은데, 이 부분은 Game에서 처리하지 않고 Players에서 해도 되지 않을까요?
값을 보관하고 처리하는 객체(하위 모듈)과 그 객체를 선언해서 사용하는 객체(상위 모듈) 관계에서
상위 모듈이 클로저로 전달해서 하위 모듈은 동작을 직접 실행할수도 있지만,
세부적인 동작을 모르고 상위 모듈이 더 간단하게 동작해야 하는 경우도 있습니다. 
지금처럼 내부 속성에 접근해서 변경하려고 하는 것까지 상위에서 할 필요는 없습니다.

`players = Players(with: numbersOfPlayers)` 형태로 한 줄로 표현할 수 있을 것 같아요.

godrm commented on CardGameApp/CardGameApp/ViewController.swift:
```diff
@@ -6,52 +6,182 @@
+    
+    private var pokerGame = PokerGame(numbersOfPlayers: .two, gameMode: .fiveCardStud)
+    private var gameMode = GameMode.sevenCardStud // segmentedControl에서 선택 받은 값에 따라 변경시키기
+    private var numbersOfPlayers = NumbersOfPlayers.four
```
PokerGame()에 인자로 넘어가는 값과 아래 gameMode, numbersOfPlayers 값이 동일해야 하는거 아닌가요?

godrm commented on CardGameApp/CardGameApp/ViewController.swift:
```diff
@@ -6,52 +6,182 @@
+    private var numbersOfPlayers = NumbersOfPlayers.four
+    // MARK: - Properties
+    // StackView
+    func cardsStackView() -> UIStackView {
```
메소드 이름 자체가 동작을 표현하지 않아서 헷갈리는 부분이 있습니다. 메소드 이름은 늘 동사로 시작하도록 표현하세요.

godrm commented on CardGameApp/CardGameApp/ViewController.swift:
```diff
@@ -6,52 +6,182 @@
+        
+        pokerGame.forEachPlayer(behavior: ){
+            player in
+            let playersCardsStack = cardsStackView()
```
이 변수는 안 쓰네요.

godrm commented on CardGameApp/CardGameApp/ViewController.swift:
```diff
@@ -6,52 +6,182 @@
+        return playerCardStack
+    }
+    
+    func makePlayersCards(of stackView: UIStackView) {
```
스위프트 API 디자인 가이드를 보면 make-를 붙이는 경우가 명시되어 있습니다. 
이 메소드는 return 타입이 없는데 make-를 붙여서 어색하네요.

godrm commented on CardGameApp/CardGameApp/Model/Players.swift:
```diff
@@ -12,14 +12,19 @@ class Players {
     
     func addCard(newCard: Card) {
         players.forEach{
             player in player.addCard(newCard: newCard)
```
카드를 하나 가져와서 모든 참가자에게 다 같은 카드를 넘기고 있는 것 같은데요?

dev-Lena replied:
아하! 수정해보겠습니다!

godrm commented on CardGameApp/CardGameApp/Model/PokerGame.swift:
```diff
@@ -0,0 +1,79 @@
+    private func distributeCards() {
+        shuffleWholeCardDeck()
+        players.addCard(newCard: dealer.giveOneCard())
+        dealer.addCard(newCard: dealer.giveOneCard())
```
위에 start에서 gameMode.setCardPlacement 만큼 반복해서 카드를 나눠주는 건가요? 
그러면 매번 shuffleWholeCardDeck() 하고 dealer에게 카드를 가져와서 추가하는 방식이네요.

dev-Lena replied:
네! 맞아요! Dealer 가 가지고 있는 CardDeck에서 카드를 가져와서 Player 들에게 카드를 나눠주는 걸로 생각했어요! 혹시 다시 생각해봐야할 부분이 있거나 수정해야 할까요??

godrm replied:
가지고 있는건 상관없는데 매번 섞을 필요는 없지 않나요?

dev-Lena replied:
아! 생각해보니 그렇네요. 그러면 그 부분은 삭제하겠습니다!

godrm reviewed (commented):
뷰에 표시않더라도 Card, Game, Player 각각 제대로 동작하는지 테스트할 수 있어야 합니다. 
뷰가 떨어져서 나오는 것은 다시 보더라도, 우선 모델 객체가 제대로 동작하는 지 확인해보세요.

dev-Lena committed:
PokerGame 클래스에서 distributeCards() 메소드에서 Dealer가 가지고 있는 CardDeck을 섞어주는 코드 삭제

dev-Lena committed:
NumbersOfPlayers enum과 GameMode enum에 메소드 아규먼트레이블명 수정

dev-Lena committed:
참여수 만큼의 Player객체를 생성하는 코드를 PokerGame 에서 Players 클래스로 이동

dev-Lena committed:
메소드 이름 수정

dev-Lena committed:
카드의 Rank가 A,J,K,Q일 때 rank.rawValue가 아닌 rank를 반환하도록 하는 조건문 추가

dev-Lena committed:
ViewController에서 Dealer의 카드 UIImage들을 추가하는 메소드 추가

dev-Lena committed:
메소드로 나눠져있던 코드를 viewDidLoad()메소드에 합침

어느 부분에서 꼬인지 파악하기 어려워 메소드로 나눠놨던 기능을 일단 viewDidLoad()메소드에 코드를 합치고 정리해 봤습니다.

dev-Lena committed:
StackView를 세팅하는 setConstraintOfView 메소드 추가

dev-Lena committed:
참가자의 이름 라벨을 만드는 makeParticipantLabel 메소드 추가

dev-Lena committed:
Card 클래스에 descripteAlphabetRank(rank:) 메소드 추가 및 Card의 description 반환 방식 변경

플레이어나 딜러가 갖고있는 카드 중의 Rank가 A,Q,J,K일 경우 카드 이미지가 출력되지 않는 문제가 있었습니다.이유는 해당 카드 이미지의 이름과 카드의 description이 매칭이 되지 않아서 였습니다. 그래서 descripteAlphabetRank(rank:) 메소드 추가해 Rank타입이 A,J,Q,K일 경우 해당 알파벳이 반환되도록 했습니다.

dev-Lena committed:
PokerGame 클래스에서 destributeCards()메소드 수정

이전 코드의 문제는 Dealer가 가지고 있는 CardDeck에서 같은 카드를 가져와서 플레이어들에게 공통으로 분배하고 있었습니다. 그래서 각 플레이어에게 각각 다른 카드를 나눠주도록 수정했습니다.

dev-Lena committed:
PokerGame 클래스에서 Start()메소드 수정

카드덱을 한 번 섞은 후 게임 모드에 맞는 카드수에 맞게 카드를 분배하도록 수정했습니다.

dev-Lena committed:
ViewController 클래스에 setPlayersCards(in gameStackView:)메소드 추가

Player의 라벨과 가지고 있는 카드 이미지를 생성해 stackView에 추가하는 메소드를 추가했습니다.

dev-Lena committed:
ViewController 클래스의 setPlayersCards(in:)메소드에서 UIImageView를 세팅해주는 부분 setCardImage(of:)메소드로 분리

dev-Lena committed:
ViewController 클래스에 Dealer가 가지고 있는 카드의 이미지를 StackView에 추가하는 addDealerCards(of:)메소드 추가

dev-Lena committed:
ViewController 클래스에 Dealer 라벨과 카드 이미지를 StackView에 추가하는 setDealersCards(in:) 메소드 추가

dev-Lena committed:
ViewController 클래스에 참가자의 라벨과 카드 이미지를 세팅하는 startPokerGame(gameMode:,numbersOfPlayers:)메소드 추가

dev-Lena committed:
ViewController에 viewDidLoad() 수정

dev-Lena committed:
ViewController에 segmentedControl 구현

게임모드(5cardStud / 7cardStud)와 플레이어 수 선택에 따라서 포커게임 판이 이에 맞춰서 변경되도록 했습니다.
선택을 변경할 때마다 참가자들의 라벨과 카드들을 전부 감싸는 gameStackView가 계속 SuperView에 쌓이는 문제가 있어서 startPokerGame(gameMode:,numbersOfPlayers)메소드가 호출될 때마다 gameStackView에 있는 subView들을 지우는 코드를 추가했습니다.

dev-Lena committed:
README.md 결과화면 스냅샷 추가

dev-Lena commented:
피드백 반영 후 수정했습니다!

1. 한 Player에게 카드이미지가 여러번 들어가는 문제 해결했습니다. 
2. 모든 Player들에게 같은 카드가 들어가는 문제를 해결했습니다. 
3. 카드의 Rank가 A,Q,J,K일 때 카드 이미지가 나오지 않는 문제를 해결했습니다.
4. SegmentedControl 구현했습니다.

godrm commented on CardGameApp/CardGameApp/Model/Card.swift:
```diff
@@ -13,7 +13,7 @@ class Card: CustomStringConvertible {
     private var rank : Rank
     var description: String {
-        return "\(suit.rawValue)\(rank.rawValue)"
+            return "\(suit.description)\(descripteAlphabetRank(rank: rank))"
```
rank.description 과 descripteAlphabetRank를 합쳐도 될 것 같네요. 참고하세요.

dev-Lena replied:
네!

godrm commented on CardGameApp/CardGameApp/ViewController.swift:
```diff
@@ -198,9 +144,9 @@ class ViewController: UIViewController {
-//        case 2: startPokerGame()// 4 명
+            //        case 0: startPokerGame()// 2 명
+            //        case 1: startPokerGame()// 3 명
+        //        case 2: startPokerGame()// 4 명
```
사용하지 않는 코드나 주석은 과감하게 지워도 됩니다.

dev-Lena replied:
네! 정리하겠습니다!

godrm commented:
다음 단계를 진행해보세요.
